# Text Classification Lab v2 - Finding Trump

- online-ds-pt-100719
- 05/15/20

- This is a modified version of https://github.com/learn-co-students/dsc-text-classification-lab-online-ds-pt-100719

## Introduction

In this lab, we'll use everything we've learned so far to build a model that can classify a text document as one of two possible classes. 

## Objectives

You will be able to:

- Perform classification using a text dataset, using sensible preprocessing, tokenization, and feature engineering scheme 
- Use scikit-learn text vectorizers to fit and transform text data into a format to be used in a ML model 



# Getting Started

~~For this lab, we'll be working with the classic **_Newsgroups Dataset_**, which is available as a training data set in `sklearn.datasets`. This dataset contains many different articles that fall into 1 of 20 possible classes. Our goal will be to build a classifier that can accurately predict the class of an article based on the features we create from the article itself!~~
## Background - Finding Trump
- In this follow-up lab you will be using the provided `finding-trump.csv` to create a text classification model to determine if a Trump came from Donald Trump or one of his staffers. 

> - During the early years of his presidency, Donald Trump refused to stop using his non-secure personal Android phone for his own use. 
- White House communications typically require presidents use a secure iPhone. 
- Consequently, for a period of time in his early presidency, Tweets from Donald Trump's twitter account were generally sent from 2 different devices:
    - Twitter for iPhone
    - Twitter for Android.
    
- During this time period (in which he was still using an Android), we can safely assume that:
    - all Tweets sent from "Twitter for Android" are Trump and that
    - all Tweets sent from the "Twitter for iPhone" are one of his staffers. 

# Your Work 

In [1]:
import nltk
from nltk.corpus import stopwords
import string
from nltk import word_tokenize, FreqDist
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import accuracy_score
from sklearn.datasets import fetch_20newsgroups
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import MultinomialNB
import pandas as pd
import numpy as np
np.random.seed(0)

## Creating the dataset

In [7]:
## Load in the finding-trump.csv 
df = pd.read_csv('finding-trump.csv')
df.head()

source                                               text  \
0  Twitter Media Studio                            https://t.co/EVAEYD1AgV   
1    Twitter for iPhone                                    HAPPY NEW YEAR!   
2    Twitter for iPhone  Our fantastic First Lady! https://t.co/6iswto4WDI   
3    Twitter for iPhone            RT @DanScavino: https://t.co/CJRPySkF1Z   
4    Twitter for iPhone  RT @SenJohnKennedy: I think Speaker Pelosi is ...   

            created_at  retweet_count  favorite_count is_retweet  \
0  01-01-2020 03:12:07          25016          108830      False   
1  01-01-2020 01:30:35          85409          576045      False   
2  01-01-2020 01:22:28          27567          132633      False   
3  01-01-2020 01:18:47          10796               0       True   
4  01-01-2020 01:17:43           8893               0       True   

                id_str  
0  1212209862094012416  
1  1212184310389850119  
2  1212182267113680896  
3  1212181341078458369  
4  1212181071988703232

### Steps for Creating the Training and Test Data
1. Create a date time index for the df from "created_at".
2. Find the date & time for the first and last tweets where "source" = `Twitter for Android` 
3. Limit the dataset to only Tweets that occurred between the first and last Twitter for Android tweets were sent.

4. Next, remove any Tweets that were sent from any device other than "Twitter for Android" and "Twitter for iPhone".

5. Create a new "target"/"trump" column where:
    - 1 = the tweet source was Twitter for Android
    - 0 = the tweet source was Twitter for iPhone
    

### Data for Modeling
- The data for the classification will come from the `text` column.
- The target will come from the "target"/"trump" column.

- Once you vectorize the dataset, you will then use train-test-split to create X_train,X_test,y_train,y_test.

In [ ]:
data = None
target = None

In [4]:
# Your code here

(2814,)

# NOTE: ALL FURTHER INSTRUCTIONS BELOW HAVE NOT BEEN UPDATED. 

- Most of the steps can be done the same way as the original lab, but you are NOT required to use the same approaches as they describe below.


# Original Lab Continued:

Our dataset contains 2,814 different articles spread across the five classes we chose. 

### Cleaning and Preprocessing Our Data

Now that we have our data, the fun part begins. We'll need to begin by preprocessing and cleaning our text data. As you've seen throughout this section, preprocessing text data is a bit more challenging that working with more traditional data types because there's no clear-cut answer for exactly what sort of preprocessing and cleaning we need to do. Before we can begin cleaning and preprocessing our text data, we need to make some decisions about things such as:

* Do we remove stop words or not?
* Do we stem or lemmatize our text data, or leave the words as is?
* Is basic tokenization enough, or do we need to support special edge cases through the use of regex?
* Do we use the entire vocabulary, or just limit the model to a subset of the most frequently used words? If so, how many?
* Do we engineer other features, such as bigrams, or POS tags, or Mutual Information Scores?
* What sort of vectorization should we use in our model? Boolean Vectorization? Count Vectorization? TF-IDF? More advanced vectorization strategies such as Word2Vec?


These are all questions that we'll need to think about pretty much anytime we begin working with text data. 

Let's get right into it. We'll start by getting a list of all of the english stopwords, and concatenating them with a list of all the punctuation. 

In the cell below:

* Get all the english stopwords from `nltk` 
* Get all of the punctuation from `string.punctuation`, and convert it to a list 
* Add the two lists together. Name the result `stopwords_list` 
* Create another list containing various types of empty strings and ellipses, such as `["''", '""', '...', '``']`. Add this to our `stopwords_list`, so that we won't have tokens that are only empty quotes and such  

In [ ]:
stopwords_list = None


Great! We'll leave these alone for now, until we're ready to remove stop words after the tokenization step. 

Next, let's try tokenizing our dataset. In order to save ourselves some time, we'll write a function to clean our dataset, and then use Python's built-in `map()` function to clean every article in the dataset at the same time. 

In the cell below, complete the `process_article()` function. This function should:

* Take in one parameter, `article` 
* Tokenize the article using the appropriate function from `nltk` 
* Lowercase every token, remove any stopwords found in `stopwords_list` from the tokenized article, and return the results 

In [ ]:
def process_tweet(article):
    pass  

Now that we have this function, let's go ahead and preprocess our data, and then move into exploring our dataset. 

In the cell below:

* Use Python's `map()` function and pass in two parameters: the `process_article` function and the `data`. Make sure to wrap the whole map statement in a `list()`.

**_Note:_** Running this cell may take a minute or two!

In [ ]:
processed_data = None

Great. Now, let's inspect the first article in `processed_data` to see how it looks. 

Do this now in the cell below.

In [ ]:
processed_data[0]

Now, let's move onto exploring the dataset a bit more. Let's start by getting the total vocabulary size of the training dataset. We can do this by creating a `set` object and then using it's `.update()` method to iteratively add each article. Since it's a set, it will only contain unique words, with no duplicates. 

In the cell below:

* Create a `set()` object called `total_vocab` 
* Iterate through each tokenized article in `processed_data` and add it to the set using the set's `.update()` method 
* Once all articles have been added, get the total number of unique words in our training set by taking the length of the set 

In [ ]:
total_vocab = None


### Exploring Data With Frequency Distributions

Great -- our processed dataset contains 46,990 unique words! 

Next, let's create a frequency distribution to see which words are used the most! 

In order to do this, we'll need to concatenate every article into a single list, and then pass this list to `FreqDist()`. 

In the cell below:

* Create an empty list called `articles_concat` 
* Iterate through `processed_data` and add every article it contains to `articles_concat` 
* Pass `articles_concat` as input to `FreqDist()`  
* Display the top 200 most used words  

In [ ]:
articles_concat = None

In [ ]:
articles_freqdist = None


At first glance, none of these words seem very informative -- for most of the words represented here, it would be tough to guess if a given word is used equally among all five classes, or is disproportionately represented among a single class. This makes sense, because this frequency distribution represents all the classes combined. This tells us that these words are probably the least important, as they are most likely words that are used across multiple classes, thereby providing our model with little actual signal as to what class they belong to. This tells us that we probably want to focus on words that appear heavily in articles from a given class, but rarely appear in articles from other classes. You may recall from previous lessons that this is exactly where **_TF-IDF Vectorization_** really shines!

### Vectorizing with TF-IDF

Although NLTK does provide functionality for vectorizing text documents with TF-IDF, we'll make use of scikit-learn's TF-IDF vectorizer, because we already have experience with it, and because it's a bit easier to use, especially when the models we'll be feeding the vectorized data into are from scikit-learn, meaning that we don't have to worry about doing any extra processing to ensure they play nicely together. 

Recall that in order to use scikit-learn's `TfidfVectorizer()`, we need to pass in the data as raw text documents -- the `TfidfVectorizer()` handles the count vectorization process on it's own, and then fits and transforms the data into TF-IDF format. 

This means that we need to:

* Import `TfidfVectorizer` from `sklearn.feature_extraction.text` and instantiate `TfidfVectorizer()` 
* Call the vectorizer object's `.fit_transform()` method and pass in our `data` as input. Store the results in `tf_idf_data_train` 
* Also create a vectorized version of our testing data, which can be found in `newsgroups_test.data`. Store the results in `tf_idf_data_test`. 


**_NOTE:_** When transforming the test data, use the `.transform()` method, not the `.fit_transform()` method, as the vectorizer has already been fit to the training data. 

In [ ]:
# Import TfidfVectorizer

In [ ]:
vectorizer = None

In [ ]:
tf_idf_data_train = None

In [ ]:
tf_idf_data_test = None

### Modeling Our Data

Great! We've now preprocessed and explored our dataset, let's take a second to see what our data looks like in vectorized form. 

In the cell below, get the shape of `tf_idf_data`.

In [16]:
# Your code here

(2814, 36622)

Our vectorized data contains 2,814 articles, with 36,622 unique words in the vocabulary. However, the vast majority of these columns for any given article will be zero, since every article only contains a small subset of the total vocabulary. Recall that vectors mostly filled with zeros are referred to as **_Sparse Vectors_**. These are extremely common when working with text data. 

Let's check out the average number of non-zero columns in the vectors. Run the cell below to calculate this average. 

In [ ]:
non_zero_cols = tf_idf_data_train.nnz / float(tf_idf_data_train.shape[0])
print("Average Number of Non-Zero Elements in Vectorized Articles: {}".format(non_zero_cols))

percent_sparse = 1 - (non_zero_cols / float(tf_idf_data_train.shape[1]))
print('Percentage of columns containing 0: {}'.format(percent_sparse))

As we can see from the output above, the average vectorized article contains 107 non-zero columns. This means that 99.7% of each vector is actually zeroes! This is one reason why it's best not to create your own vectorizers, and rely on professional packages such as scikit-learn and NLTK instead -- they contain many speed and memory optimizations specifically for dealing with sparse vectors. This way, we aren't wasting a giant chunk of memory on a vectorized dataset that only has valid information in 0.3% of it. 

Now that we've vectorized our dataset, let's create some models and fit them to our vectorized training data. 

In the cell below:

* Instantiate `MultinomialNB()` and `RandomForestClassifier()`. For random forest, set `n_estimators` to `100`. Don't worry about tweaking any of the other parameters  
* Fit each to our vectorized training data 
* Create predictions for our training and test sets
* Calculate the `accuracy_score()` for both the training and test sets (you'll find our training labels stored within the variable `target`, and the test labels stored within `newsgroups_test.target`) 

In [ ]:
nb_classifier = None
rf_classifier = None

In [ ]:

nb_train_preds = None
nb_test_preds = None

In [ ]:

rf_train_preds = None
rf_test_preds = None

In [ ]:
nb_train_score = None
nb_test_score = None
rf_train_score = None
rf_test_score = None

print("Multinomial Naive Bayes")
print("Training Accuracy: {:.4} \t\t Testing Accuracy: {:.4}".format(nb_train_score, nb_test_score))
print("")
print('-'*70)
print("")
print('Random Forest')
print("Training Accuracy: {:.4} \t\t Testing Accuracy: {:.4}".format(rf_train_score, rf_test_score))

### Interpreting Results

**_Question:_** Interpret the results seen above. How well did the models do? How do they compare to random guessing? How would you describe the quality of the model fit?

Write your answer below:

In [ ]:
# Your answer here

# Summary

In this lab, we used our NLP skills to clean, preprocess, explore, and fit models to text data for classification. This wasn't easy -- great job!!